tests/phpunit/integration/api/html/beansWrapMarkup.php

```diff
@@ -0,0 +1,173 @@
+ *
+ * @since   1.5.0
+ */
+
+namespace Beans\Framework\Tests\Integration\API\HTML;
+
+use Beans\Framework\Tests\Integration\API\HTML\Includes\HTML_Test_Case;
+use Brain\Monkey;
+
+require_once __DIR__ . '/includes/class-html-test-case.php';
+
+/**
+ * Class Tests_BeansWrapMarkup
+ *
+ * @package Beans\Framework\Tests\Integration\API\HTML
+ * @group   api
+ * @group   api-html
+ */
+class Tests_BeansWrapMarkup extends HTML_Test_Case {
+
+	/**
+	 * Since we do not have direct access to the instance, we need to get it from WordPress.
+	 *
+	 * @since 1.5.0
+	 *
+	 * @param string $hook     To given hook's event name.
+	 * @param int    $priority The priority number for the callback.
+	 *
+	 * @return \_Beans_Anonymous_Actions
+	 */
+	protected function get_instance_from_wp( $hook, $priority ) {
+		global $wp_filter;
+
+		// Run a couple of checks to make sure WordPress has the registration.
+		$this->assertArrayHasKey( $hook, $wp_filter );
+		$this->assertArrayHasKey( $priority, $wp_filter[ $hook ] );
+
+		return end( $wp_filter[ $hook ]->callbacks[ $priority ] )['function'][0];
+	}
+
+	/**
+	 * Test beans_wrap_markup() should register beans_open_markup() to the given ID's '_before_markup' hook.
+	 */
+	public function test_should_register_beans_open_markup_to_given_id_before_markup_hook() {
+		$this->assertTrue( beans_wrap_markup( 'foo', 'new_foo', 'div', array( 'class' => 'test-wrap' ) ) );
+
+		// Check that it did register a callback to the hook.
+		$this->assertTrue( has_action( 'foo_before_markup' ) );
+
+		// Check that the correct arguments were stored in the instance.
+		$instance = $this->get_instance_from_wp( 'foo_before_markup', 9999 );
+		$this->assertSame(
+			array(
+				'beans_open_markup',
+				array(
+					1 => 'new_foo',
+					2 => 'div',
+					3 => array( 'class' => 'test-wrap' ),
+				),
+			),
+			$instance->callback
+		);
+
+		// Clean up.
+		remove_action( 'foo_before_markup', array( $instance, 'callback' ), 9999 );
+	}
+
+	/**
+	 * Test beans_wrap_markup() should register beans_close_markup() to the given ID's '_after_markup' hook.
+	 */
+	public function test_should_register_beans_close_markup_to_given_id_after_markup_hook() {
+		$this->assertTrue( beans_wrap_markup( 'foo', 'new_foo', 'div', array( 'class' => 'test-wrap' ) ) );
+
+		// Check that it did register a callback to the hook.
+		$this->assertTrue( has_action( 'foo_after_markup' ) );
+
+		// Check that the correct arguments were stored in the instance.
+		$instance = $this->get_instance_from_wp( 'foo_after_markup', 1 );
+		$this->assertSame(
+			array(
+				'beans_close_markup',
+				array(
+					1 => 'new_foo',
+					2 => 'div',
+				),
+			),
+			$instance->callback
+		);
+
+		// Clean up.
+		remove_action( 'foo_after_markup', array( $instance, 'callback' ), 1 );
+	}
+
+	/**
+	 * Test beans_wrap_markup() should not pass the given attributes to anonymous action.
+	 */
+	public function test_should_not_pass_attributes_for_after_markup_hook() {
```

Comment:
>Don't the previous and last test cover that no attributes are passed to the after markup hook?

Nope, the previous test focused on the registration process only, i.e. by checking if the anonymous action's callback did get registered to the `'foo_after_markup'` hook.

This test is checking specifically that the attributes are not being passed to the anonymous action.  How?  It compares what was stored in the anonymous action mock's property "callback" comparing it to what we expect to be stored:

```
$this->assertSame( $args, $anonymous_action_mock->callback );
```

>By anonymous action you mean $new_id was not passed
Nope, I mean the anonymous action.  Look at the code:

```
function beans_wrap_markup( $id, $new_id, $tag, $attributes = array() ) {
	$args = func_get_args();
	unset( $args[0] );

	_beans_add_anonymous_action( $id . '_before_markup', array( 'beans_open_markup', $args ), 9999 );

	unset( $args[3] );

	_beans_add_anonymous_action( $id . '_after_markup', array( 'beans_close_markup', $args ), 1 );

	return true;
}
```

Notice that it is calling the function `_beans_add_anonymous_action()`.  That function creates an instance of the anonymous action.  

In the test suite, I'm mocking (or stubbing) that functionality.

But the terminology within Beans is: "anonymous action".

Reply:
The unit test is great. It makes total sense.

To me, the integration test doesn't explain itself clearly enough.

- The only reference to anonymous action is in the return parameter DocBlock of `get_instance_from_wp()` about 70 lines above. 

- There are only two differences to the previous test:
 1. $new_id is an empty string
 2. No check if the callback was registered ($this->assertTrue( has_action( 'foo_after_markup' ) );

- The term `anonymous action` is nowhere to be found in the test. 
My brain tries to fill the void by finding the most obvious difference -> an empty string passed -> missing parameter == anonymous action.

Reply:
I see the confusion now.  The test `test_should_register_beans_close_markup_to_given_id_after_markup_hook` had a check in it for the callback property.  That's wrong.

To improve readability, I moved the `wp_filter` tests into each integration method.

>The term anonymous action is nowhere to be found in the test.

Correct. But when you look at the function's code, it's used multiple times.  The function is calling another Beans' function that creates the anonymous function.  It is the `$instance` that we are checking.  Would you prefer `$instance` to be renamed to `$anonymous_action`?  Would that help?

Reply:
Okay, here's what I did:

1. I renamed the `$instance` to `$anonymous_action` to better represent what the object is.
2. The intent of `test_should_not_pass_attributes_for_after_markup_hook` test is to check that the `unset( $args[3] );` in the `beans_wrap_markup()` function works as intended.  Therefore, I simplified the test to _only_ check that the attributes do not exist in the anonomyous action, i.e.meaning they were not stored in the `$callback` property.
3. I moved the `$wp_filter` tests down into the appropriate methods so that you can see them instead of having to scroll up to them.

Let me know if that provides better clarity.

Reply:
Thank you.
Sorry for not being clearer in my initial comments. 

1. If we want to point out that it is an instance, `$anonymous_action_instance` is longer but  makes sense to me. Otherwise, it's great. 
2. I think this was the main point of my `why is the test here?`  I like tests that will tell us clearly what we broke.
3. That's an added bonus :) If you'd rather keep it in the `get_instance_from_wp()` method to be more DRY, that's fine by me.

Reply:
Not a problem.  Your points were spot on and valid.  It should took me a couple of iterations to figure out how to achieve what you were looking for.

1. I think `$anonymous_action` is fine.  It's an object.  I don't think we need to go so far as to declare that it's one.  We are using it as an object within each test.  Therefore, it should be clear as to its data type.
2. Cool.  I didn't want us to loose the fact that we do the unset.  It can get lost in the details.  Should we break that down the road, it might not be clear why without that test.
3. True. But you're right.  It's more clear now.  Readability is important.

Thanks for your help on this one!  You did a great job!